### PR TITLE
Harry/localhost fix

### PR DESCRIPTION
**Pull request by hwh33**
Depends on https://github.com/getlantern/netx/pull/16.

The issue with the current code is that `Resolve("tcp", "localhost")` will make a random choice between "127.0.0.1" or "[::1]".  AFAICT, most applications listening on localhost are only serving one of those addresses, almost always the IPv4 address.  Crucially, a Go server told to listen on localhost will only serve 127.0.0.1.

This is one of the issues plaguing lantern-desktop's `desktop/TestProxying`.  Protocols which use `netx.Resolve` occasionally resolve `localhost` to `[::1]` and cannot reach the local proxy server.

Technically, this change might be wrong for a server listening on `[::1]`.  Really, `Resolve` should probably return all resolved IP addresses and let the caller choose the right one for their needs.  This way, dialers could try IPv4 first, then move on to IPv6.  I believe this is what Go's `net` package does.  I think returning the IPv4 address for localhost should be okay though.  We are unlikely to run into local servers running exclusively over IPv6.

### CODE REVIEW

**Comment by hwh33** [comment]
This will include #16's changes until that PR is merged.  To see this PR's changes in isolation, please look at https://github.com/getlantern/netx/pull/17/commits/c312af0111694e07dca8db0aa72c3360470495f0

**Comment by oxtoacart** [comment]
> Really, `Resolve` should probably return all resolved IP addresses and let the caller choose the right one for their needs. This way, dialers could try IPv4 first, then move on to IPv6

I strongly agree with this.

**Comment by hwh33** [comment]
> Do we only ever get into this code path resolving Go servers?

No, and that's a big part of why this change is not the 100% correct way to go.  IMO, an API change is the only 100% correct solution.  More on this below.

>> Really, Resolve should probably return all resolved IP addresses and let the caller choose the right one for their needs. This way, dialers could try IPv4 first, then move on to IPv6

> I strongly agree with this.

I could instead create a `netx/v2` which makes the suggested API change.  I don't think we can just supplement existing API since, as discussed above, the existing API is inherently flawed.  We'd also have to make sure dependents are unilaterally changed to `netx/v2` as we rely on the ability to change the global resolve function.

Depends on how much we feel like my proposed fix is itself a problem.  I'll admit that I don't love it.  It could potentially lead to a subtle failure in which we fail to resolve/dial a localhost listener only serving IPv6.  This seems like an unlikely scenario though.

What do you think?

**Comment by oxtoacart** [comment]
@hwh33 if the primary use case here is our own code connecting to localhost, can we just change our code to connect to 127.0.0.1?

**Comment by hwh33** [comment]
> @hwh33 if the primary use case here is our own code connecting to localhost, can we just change our code to connect to 127.0.0.1?

That was the first thing I tried actually.  The immediate issue was that the certs specified 'localhost'.  We can of course create new certs valid for 127.0.0.1, but that made me realize that the issue was really netx's behavior.  I think ignoring this issue is just passing the buck to someone else.  Any test for a protocol using netx.Resolve or netx.Dial and targeting localhost is going to see random failures.

So, while I don't love my proposed fix, I think doing nothing is worse.

**Comment by hwh33** [comment]
Looks like the following repos all use netx.  I think almost all of these use netx directly, but some may be indirect users (netx only appears in their go.sum):

https://github.com/getlantern/vain-haproxy-dns
https://github.com/getlantern/dnsgrab
https://github.com/getlantern/fronted
https://github.com/getlantern/packetforward
https://github.com/getlantern/idletiming
https://github.com/getlantern/cmux
https://github.com/getlantern/detour
https://github.com/getlantern/lampshade
https://github.com/getlantern/tlsproxy
https://github.com/getlantern/proxy
https://github.com/getlantern/tlsdialer
https://github.com/getlantern/tinywss
https://github.com/getlantern/quicwrapper
https://github.com/getlantern/http-proxy-lantern
https://github.com/getlantern/http-proxy
https://github.com/getlantern/kcpwrapper
https://github.com/getlantern/android-lantern
https://github.com/getlantern/mitm
https://github.com/getlantern/flashlight
https://github.com/getlantern/genconfig
https://github.com/getlantern/proxybench
https://github.com/getlantern/checkfallbacks
https://github.com/getlantern/replica
https://github.com/getlantern/lantern-desktop
https://github.com/getlantern/bbrconn
https://github.com/getlantern/enhttp

These repos also use netx, but I believe these might be inactive:

https://github.com/getlantern/lanternlib
https://github.com/getlantern/quic0
https://github.com/getlantern/interceptor

**Comment by oxtoacart** [comment]
> I think ignoring this issue is just passing the buck to someone else

IIUC there's 2 distinct questions.

1. Should the resolve function in netx respect the network type? Clearly the answer to that is yes, and your already merged PR fixes that.

2. Should netx assume that someone dialing "localhost" always intends to dial with IPv4 when they weren't specific about it? IMHO the answer to that is no, and whoever is doing the dialing should either be specific about the protocol, specific about the IP, or be willing to live with getting an IPv6 address if one is available.

**Comment by hwh33** [comment]
> Should netx assume that someone dialing "localhost" always intends to dial with IPv4 when they weren't specific about it? IMHO the answer to that is no, and whoever is doing the dialing should either be specific about the protocol, specific about the IP, or be willing to live with getting an IPv6 address if one is available.

Hmm I do get why you don't like the resolve function assuming the caller wants to dial IPv4.  I don't like that either.  _However_, I think `Resolve` answers a question like, "Can you give me an address at which I can reach `somehost`?"  In the case of "localhost", the answer AFAICT is almost never "[::1]" and is almost always "127.0.0.1".  So returning "[::1]" half the time is tantamount to giving the wrong answer half the time.

Compare this to querying `Resolve` for say, "google.com".  It returns a random selection from all possible answers, but that's okay because all of those choices are correct.  In the case of "localhost", "[::1]" is probably not going to connect you to the server.  If `Resolve` instead answered a question like, "Can you give me an address which is technically supposed to serve `somehost`?" then I would agree with you.

What do you think about that?  Do you still think the proposed fix is wrong?  If so, what do you propose doing instead?

**Comment by oxtoacart** [comment]
@hwh33 To get my head around this, I wrote a little Go program:

```
package main

import (
	"fmt"
	"net"
)

func main() {
	ip, _ := net.ResolveIPAddr("ip", "localhost")
	fmt.Printf("ip: %v\n", ip)
	ip4, _ := net.ResolveIPAddr("ip4", "localhost")
	fmt.Printf("ip4: %v\n", ip4)
	ip6, _ := net.ResolveIPAddr("ip6", "localhost")
	fmt.Printf("ip6: %v\n", ip6)
}
```

I tried running this on both a Mac and an Ubuntu box, both of which have IPv6 enabled on the loopback interface. Here's what I observed.

1. Using the plain network `ip` always yields an IPv4 address
2. On Ubuntu, the `ip6` network actually yields nothing?!
3. I could not figure out a way to turn off IPv4 on the loopback interface to test Go's behavior, so in practice I think it's safe to assume localhost will always have at least an IPv4 address

I then wrote a program to resolve `www.google.com`

```
package main

import (
	"fmt"
	"net"
)

func main() {
	ip, _ := net.ResolveIPAddr("ip", "www.google.com")
	fmt.Printf("ip: %v\n", ip)
	ip4, _ := net.ResolveIPAddr("ip4", "www.google.com")
	fmt.Printf("ip4: %v\n", ip4)
	ip6, _ := net.ResolveIPAddr("ip6", "www.google.com")
	fmt.Printf("ip6: %v\n", ip6)
}
```

With this, I observed that:

1. The plain network `ip` always yields an IPv4 address even when an IPv6 address is available
2. If I disable IPv4 on my Mac, the plain network `ip` still always yields an IPv4 address (I even tried flushing my Mac's DNS resolver cache, same result)

So long story short, I think it's okay to always return an IPv4 address for localhost. Furthermore, it's probably safe (maybe even good) to have resolve preferentially return IPv4 addresses for unqualified lookups. In other words, if someone is using `ip` or `tcp` or `udp` and name resolution yields both IPv4 and IPv6 addresses, we should return one of the IPv4 addresses. Go seems even more aggressive than this by always returning IPv4, but I'm not sure I'm ready to go that far.

**Comment by hwh33** [comment]
Nice, smart to look to `net.ResolveIPAddr`!  I got stuck looking at `net.LookupIP` (which is what netx uses) and didn't even notice `net.ResolveIPAddr`.

> In other words, if someone is using ip or tcp or udp and name resolution yields both IPv4 and IPv6 addresses, we should return one of the IPv4 addresses.

I like this a lot.  This is much cleaner than my approach and solves the problem (and maybe then some).  Thanks a bunch for the thorough review so far and thoughtfulness w.r.t. the solution.  This change could have some subtle implications down the road, so I'm glad to have something that actually feels correct.  I'll close this PR and open a new one with your suggested solution.

Also this was funny and surprising!

> On Ubuntu, the ip6 network actually yields nothing?!

**Comment by hwh33** [comment]
New PR: https://github.com/getlantern/netx/pull/19

**Comment by Crosse** [comment]
I know this one is closed, but I wanted to shed some light on @oxtoacart's [comment](https://github.com/getlantern/netx/pull/17#issuecomment-983270280) about Ubuntu not returning anything for an IPv6 lookup of `localhost`.

When doing name lookups, `/etc/hosts` is usually consulted first (as `files` is usually the first thing listed in `/etc/nsswitch.conf` for the `hosts` database). Since that file is where entries for `localhost` will come from, any sort of name resolution will hinge on what is in `/etc/hosts`. On CentOS, for instance, `/etc/hosts` looks like this on a machine with the hostname "mirror":

```
$ grep -v '^#' /etc/hosts
127.0.1.1       mirror
127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4
::1         localhost localhost.localdomain localhost6 localhost6.localdomain6
```

Running the example program above gives this:

```
$ go run main.go
ip: 127.0.0.1
ip4: 127.0.0.1
ip6: ::1
```

Contrast that with Ubuntu 20.04, where `/etc/hosts` looks like this:

```
$ cat /etc/hosts
127.0.0.1 localhost

# The following lines are desirable for IPv6 capable hosts
::1 ip6-localhost ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
ff02::3 ip6-allhosts
```

Since `::1` is listed in the file, but specifically does _not_ resolve to `localhost`, `net.ResolveIPAddr()` rightfully returns nothing for it. (And if you tacked `localhost` onto the existing line for `::1`, `net.ResolveIPAddr()` with then return a value for `ip6`.)

**Comment by oxtoacart** [comment]
Thanks @Crosse!

**Comment by hwh33** [comment]
Very interesting.  Thanks @Crosse!
